helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java

```diff
@@ -102,15 +114,22 @@ public void process(ClusterEvent event) throws Exception {
 
   // update all pending messages to CurrentStateOutput.
   private void updatePendingMessages(LiveInstance instance, Collection<Message> pendingMessages,
-      CurrentStateOutput currentStateOutput, Collection<Message> pendingRelayMessages,
-      Map<String, Resource> resourceMap) {
+      Collection<Message> pendingRelayMessages,
+      Map<String, Map<String, Message>> existingStaleMessages,
+      CurrentStateOutput currentStateOutput, Map<String, Resource> resourceMap) {
     String instanceName = instance.getInstanceName();
     String instanceSessionId = instance.getEphemeralOwner();
 
     // update all pending messages
     for (Message message : pendingMessages) {
+      // ignore existing stale messages
+      if (existingStaleMessages.getOrDefault(instanceName, Collections.emptyMap())
```

Comment:
We can call currentStateOutput.getStaleMessagesByInstance() to get the message set as the input, right?

Reply:
CurrentStateOutput currently does not have stale messages in it. It's not necessary because every class has access to cache.